cmd/oras/cp.go

```diff
@@ -131,34 +137,43 @@ func runCopy(opts copyOptions) error {
 	}
 
 	var desc ocispec.Descriptor
-	if ref := opts.To.Reference; ref == "" {
-		// push to the destination with digest only if no tag specified
-		desc, err = src.Resolve(ctx, opts.From.Reference)
+	if ref := opts.To.Reference; ref == "" || opts.recursive {
+		// need resolving first
+		rOpts := oras.DefaultResolveOptions
+		rOpts.TargetPlatform = opts.Platform.Platform
+		desc, err = oras.Resolve(ctx, src, opts.From.Reference, rOpts)
 		if err != nil {
 			return err
 		}
 		if opts.recursive {
+			// not ExtendCopy since we already have desc
 			err = oras.ExtendedCopyGraph(ctx, src, dst, desc, extendedCopyOptions.ExtendedCopyGraphOptions)
 		} else {
 			err = oras.CopyGraph(ctx, src, dst, desc, extendedCopyOptions.CopyGraphOptions)
 		}
+		if err != nil {
+			return err
+		}
+		if ref != "" {
+			err = dst.Tag(ctx, desc, ref)
+		}
```

Comment:
By decomposing `ExtendedCopy` to `Resolve`-`ExtendedCopyGraph`-`Tag`, we lose the optimization for `FetchReference` and `PushReference` when platform is not specified, which is the most case.

Reply:
Yes, you are right. Updated to only do `Resolve`-`ExtendedCopyGraph`-`Tag` when recursively copying with platform specified.